### ISSUE TITLE
Building SVF as a shared library

### ISSUE THREAD

**Issue by FrancisYL** [SVF-tools/SVF#156]
Hi, first thank you for sharing and maintaining such a great project!

I'm trying to write my own LLVM pass and declare WPAPass as a dependency so as to get alias results of pointers and memory locations. In order to declare WPAPass as a dependency, I need to build a shared library of SVF and load it using `opt`.

SVF is built as a static library by default. I tried to build SVF as a shared library for my project but I encountered the problem described in #53. Even after I removed `${llvm_libs}`, I got segmentation fault querying alias results, which makes me think that SVF is not designed to be used as a shared library at its design phase.

So my questions are, first, what's the initial design reasoning for SVF, why building it as a static library instead of a shared library? Second, is there any way I can build SVF as a shared library so that I can load and declare WPAPass as my dependency to use AliasAnalysis? Thank you!

**Comment by yuleisui** [comment]
@HongxuChen would you mind taking a look at the issue, which is similar to the one you addressed before?

**Comment by HongxuChen** [comment]
@FrancisYL Can you provide the segfault backtrace and the triggering example to describe your issues?
For the shared library issue, do you mean to use it as a plugin pass (which can be used by utilities such as `opt -load xxx.so`)? I remember SVF previously supported this but due to some clashes the support was dropped.

**Comment by FrancisYL** [comment]
@HongxuChen Here are the changes I made.

First, for `lib/CMakeLists.txt`, I changed `add_llvm_library(Svf ${SOURCES})` to `add_llvm_library(Svf SHARED ${SOURCES})` and commented out all the occurance of `target_link_libraries(Svf ${llvm_libs})`. This successfully build SVF and generate `libSvf.so`.

I then wrote my own module pass (called CAT) calling WPAPass as my dependency by adding `analysisUsage.addRequired<WPAPass>()`. In runOnModule method I fetched the analysis result `WPAPass& wpa = getAnalysis<WPAPass>()` and provide two memory location to check whether they alias `wpa.alias(MemoryLocation::get(memInst), MemoryLocation::get(memInst2))`

I used `opt` to call my pass using
`opt -load ~/projects/SVF-9/Release-build/lib/libSvf.so -load ~/CAT/lib/CAT.so program.bc -disable-output -CAT`. I believed the library is successfully loaded and there's no 'bitcode-mdindex-threshold' is registered more than once issue mentioned in #53.

But this time I got a Segmentation fault when querying `WPAPass::alias`.
```
Stack dump:
0.	Program arguments: opt -load /home/yso0488/projects/SVF-9/Release-build/lib/libSvf.so -load /home/yso0488/CAT/lib/CAT.so program.bc -disable-output -CAT 
1.	Running pass 'Simple Code Analysis and Transformation Pass' on module 'program.bc'.
#0 0x00000000018bcd2a llvm::sys::PrintStackTrace(llvm::raw_ostream&) /project/llvm/9.0.0/src/lib/Support/Unix/Signals.inc:537:3
#1 0x00000000018babb4 llvm::sys::RunSignalHandlers() /project/llvm/9.0.0/src/lib/Support/Signals.cpp:68:20
#2 0x00000000018bad15 SignalHandler(int) /project/llvm/9.0.0/src/lib/Support/Unix/Signals.inc:385:1
#3 0x00007fb083317630 __restore_rt (/lib64/libpthread.so.0+0xf630)
#4 0x00007fb081c17de4 WPAPass::alias(llvm::Value const*, llvm::Value const*) (/home/yso0488/projects/SVF-9/Release-build/lib/libSvf.so+0x19ede4)
make: *** [Makefile:15: program] Segmentation fault
make: *** Deleting file 'program'
```

**Comment by HongxuChen** [comment]
@FrancisYL Maybe try debug build and provide the gdb backtrace (excluding CAT.so)?
But I suggest not using svf.so as a plugin pass since there are probably other issues (I met one but cannot  remember exactly) and the results may not be correctly chained for other passes.

**Comment by FrancisYL** [comment]
@HongxuChen 

This is the gdb backtrace I got through Release-build
```
(gdb) backtrace
#0  0x00007ffff63680e0 in PAG::hasValueNode (this=0x0, V=0x26aa4a8) at /home/yso0488/projects/SVF-9/include/MemoryModel/PAG.h:405
#1  0x00007ffff63fb155 in WPAPass::alias (this=0x268db30, V1=0x26aa4a8, V2=0x26aa4a8) at /home/yso0488/projects/SVF-9/lib/WPA/WPAPass.cpp:196
#2  0x00007ffff63fc085 in WPAPass::alias (this=0x268db30, LocA=..., LocB=...) at /home/yso0488/projects/SVF-9/include/WPA/WPAPass.h:85
#3  0x00007ffff5c376c0 in (anonymous namespace)::CAT::checkFunction (this=0x268db00, M=..., F=..., wpa=...) at /home/yso0488/projects/CAT/CAT-c/catpass/CatPass.cpp:107
#4  0x00007ffff5c37051 in (anonymous namespace)::CAT::runOnModule (this=0x268db00, M=...) at /home/yso0488/projects/CAT/CAT-c/catpass/CatPass.cpp:38
#5  0x0000000000fe14fa in llvm::legacy::PassManagerImpl::run(llvm::Module&) ()
#6  0x00000000006c3503 in main ()
```
It looks like that PAG::hasValueNode itself leads to the segmentation fault. I wondered if its' casued by removing `target_link_libraries(Svf ${llvm_libs})`.

I'm recently trying to build my `CAT` pass linking with `SVF` so instead of loading a `libSVF.so` I can initiate a WPAPass directly inside my code. I'm not sure whether this approach is a good idea but I'm still interested in the underlying design logic behind `SVF`, that is, why building a static library instead of a shared library? @yuleisui Could you comment on that?

Thanks!

**Comment by HongxuChen** [comment]
@FrancisYL It's probably not due to removing linking dependencies. I suppose it's lacking some prologue code that a regular SVF executes; you may refer to [tools/WPA/WPA.cpp](https://github.com/SVF-tools/SVF/blob/master/tools/WPA/wpa.cpp#L47-L50) for help. Additionally, I suggest not loading your own pass when testing functionalities of SVF.

**Comment by pzieris** [comment]
Hi everyone, as luck would have it I stumbled on this great project a few days ago when looking for a MemorySSA more precise than the one LLVM provides. As I am currently writing a LLVM pass as well, I looked into the problem @FrancisYL describes in this thread.

Taking @HongxuChen's suggestion into account: Calling `opt -load libSvf.so -load MyPass.so -ander -mypass ...` solved the problem for me. I guess without this option, the WPA pass hardly does anything and the PAG object never gets initialized. Hence, the segmentation fault when trying to call `hasValueNode()`. I have not yet looked at the AA results whether everything is working as intended, but I hope this helps!
